From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Greg Price <greg@zulip.com>
Date: Tue, 7 Jun 2022 20:57:09 -0700
Subject: [tsflower] Replace interface-extends with spread for non-interface
 bases
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

In TypeScript, `interface T { … }` and `type T = { … }`, when there
are no other type definitions for T, mean exactly the same thing.

In Flow, they're different -- an interface is a different kind of
type from an object type.  In particular, a further interface
definition can extend an interface, but can't extend an object type.

Currently, TsFlower emits an interface when it sees an interface.
That's good if the type definitions are going to go on to use that
interface as the base for another interface, or for a class; it's
bad if it was extending another type that turns out to be just an
object type.

Probably an improvement to the heuristic would be to convert an
interface declaration to an object type, with `extends` turned into
a spread, when the base is an object type -- as it is in these
examples.
---
 .../PagerViewNativeComponent.js.flow          | 30 ++++++++++---------
 .../lib/typescript/src/SafeArea.types.js.flow | 24 ++++++++-------
 .../typescript/src/SafeAreaContext.js.flow    | 12 ++++----
 .../src/specs/NativeSafeAreaProvider.js.flow  |  8 +++--
 .../src/specs/NativeSafeAreaView.js.flow      | 10 ++++---
 5 files changed, 48 insertions(+), 36 deletions(-)

diff --git types/react-native-pager-view/lib/typescript/PagerViewNativeComponent.js.flow types/react-native-pager-view/lib/typescript/PagerViewNativeComponent.js.flow
index 06874a9bd..30a7c61e1 100644
--- types/react-native-pager-view/lib/typescript/PagerViewNativeComponent.js.flow
+++ types/react-native-pager-view/lib/typescript/PagerViewNativeComponent.js.flow
@@ -24,20 +24,22 @@ export type OnPageScrollStateChangedEventData = Readonly<{
   ...
 }>;
 
-export interface NativeProps extends $tsflower_subst$RN$ViewProps {
-  scrollEnabled?: WithDefault<boolean, true>;
-  layoutDirection?: WithDefault<'ltr' | 'rtl', 'ltr'>;
-  initialPage?: Int32;
-  orientation?: WithDefault<'horizontal' | 'vertical', 'horizontal'>;
-  offscreenPageLimit?: Int32;
-  pageMargin?: Int32;
-  overScrollMode?: WithDefault<'auto' | 'always' | 'never', 'auto'>;
-  overdrag?: WithDefault<boolean, false>;
-  keyboardDismissMode?: WithDefault<'none' | 'on-drag', 'none'>;
-  onPageScroll?: DirectEventHandler<OnPageScrollEventData>;
-  onPageSelected?: DirectEventHandler<OnPageSelectedEventData>;
-  onPageScrollStateChanged?: DirectEventHandler<OnPageScrollStateChangedEventData>;
-}
+export type NativeProps = {
+  ...$tsflower_subst$RN$ViewProps,
+  scrollEnabled?: WithDefault<boolean, true>,
+  layoutDirection?: WithDefault<'ltr' | 'rtl', 'ltr'>,
+  initialPage?: Int32,
+  orientation?: WithDefault<'horizontal' | 'vertical', 'horizontal'>,
+  offscreenPageLimit?: Int32,
+  pageMargin?: Int32,
+  overScrollMode?: WithDefault<'auto' | 'always' | 'never', 'auto'>,
+  overdrag?: WithDefault<boolean, false>,
+  keyboardDismissMode?: WithDefault<'none' | 'on-drag', 'none'>,
+  onPageScroll?: DirectEventHandler<OnPageScrollEventData>,
+  onPageSelected?: DirectEventHandler<OnPageSelectedEventData>,
+  onPageScrollStateChanged?: DirectEventHandler<OnPageScrollStateChangedEventData>,
+  ...
+};
 
 type PagerViewViewType = HostComponent<NativeProps>;
 
diff --git types/react-native-safe-area-context/lib/typescript/src/SafeArea.types.js.flow types/react-native-safe-area-context/lib/typescript/src/SafeArea.types.js.flow
index 3a71dca37..81a9fb120 100644
--- types/react-native-safe-area-context/lib/typescript/src/SafeArea.types.js.flow
+++ types/react-native-safe-area-context/lib/typescript/src/SafeArea.types.js.flow
@@ -40,15 +40,19 @@ export interface Metrics {
 export type InsetChangedEvent = $tsflower_subst$RN$NativeSyntheticEvent<Metrics>;
 export type InsetChangeNativeCallback = (event: InsetChangedEvent) => void;
 
-export interface NativeSafeAreaProviderProps extends $tsflower_subst$RN$ViewProps {
-  children?: $tsflower_subst$React$ReactNode;
-  onInsetsChange: InsetChangeNativeCallback;
-}
-
-export interface NativeSafeAreaViewProps extends $tsflower_subst$RN$ViewProps {
-  children?: $tsflower_subst$React$ReactNode;
-  mode?: 'padding' | 'margin';
-  edges?: Edges;
-}
+export type NativeSafeAreaProviderProps = {
+  ...$tsflower_subst$RN$ViewProps,
+  children?: $tsflower_subst$React$ReactNode,
+  onInsetsChange: InsetChangeNativeCallback,
+  ...
+};
+
+export type NativeSafeAreaViewProps = {
+  ...$tsflower_subst$RN$ViewProps,
+  children?: $tsflower_subst$React$ReactNode,
+  mode?: 'padding' | 'margin',
+  edges?: Edges,
+  ...
+};
 
 export type NativeSafeAreaViewInstance = InstanceType<typeof NativeSafeAreaView>;
diff --git types/react-native-safe-area-context/lib/typescript/src/SafeAreaContext.js.flow types/react-native-safe-area-context/lib/typescript/src/SafeAreaContext.js.flow
index 7231a733b..5e7699444 100644
--- types/react-native-safe-area-context/lib/typescript/src/SafeAreaContext.js.flow
+++ types/react-native-safe-area-context/lib/typescript/src/SafeAreaContext.js.flow
@@ -18,11 +18,13 @@ import { type EdgeInsets, type Metrics, type Rect } from './SafeArea.types';
 declare export var SafeAreaInsetsContext: $tsflower_subst$React$Context<EdgeInsets | null>;
 declare export var SafeAreaFrameContext: $tsflower_subst$React$Context<Rect | null>;
 
-export interface SafeAreaProviderProps extends $tsflower_subst$RN$ViewProps {
-  children?: $tsflower_subst$React$ReactNode;
-  initialMetrics?: Metrics | null;
-  initialSafeAreaInsets?: EdgeInsets | null;
-}
+export type SafeAreaProviderProps = {
+  ...$tsflower_subst$RN$ViewProps,
+  children?: $tsflower_subst$React$ReactNode,
+  initialMetrics?: Metrics | null,
+  initialSafeAreaInsets?: EdgeInsets | null,
+  ...
+};
 
 declare export function SafeAreaProvider(SafeAreaProviderProps): $tsflower_subst$React$JSX$Element;
 declare export function useSafeAreaInsets(): EdgeInsets;
diff --git types/react-native-safe-area-context/lib/typescript/src/specs/NativeSafeAreaProvider.js.flow types/react-native-safe-area-context/lib/typescript/src/specs/NativeSafeAreaProvider.js.flow
index 7a1594b3b..5d9e21df4 100644
--- types/react-native-safe-area-context/lib/typescript/src/specs/NativeSafeAreaProvider.js.flow
+++ types/react-native-safe-area-context/lib/typescript/src/specs/NativeSafeAreaProvider.js.flow
@@ -27,8 +27,10 @@ export type Event = Readonly<{
   ...
 }>;
 
-export interface NativeProps extends $tsflower_subst$RN$ViewProps {
-  onInsetsChange?: DirectEventHandler<Event, 'paperInsetsChange'>;
-}
+export type NativeProps = {
+  ...$tsflower_subst$RN$ViewProps,
+  onInsetsChange?: DirectEventHandler<Event, 'paperInsetsChange'>,
+  ...
+};
 declare var _default: HostComponent<NativeProps>;
 export default _default;
diff --git types/react-native-safe-area-context/lib/typescript/src/specs/NativeSafeAreaView.js.flow types/react-native-safe-area-context/lib/typescript/src/specs/NativeSafeAreaView.js.flow
index e30582cb5..87fdeaef0 100644
--- types/react-native-safe-area-context/lib/typescript/src/specs/NativeSafeAreaView.js.flow
+++ types/react-native-safe-area-context/lib/typescript/src/specs/NativeSafeAreaView.js.flow
@@ -6,16 +6,18 @@ import type { Readonly } from 'tsflower/subst/lib';
 import { typeof WithDefault } from 'react-native/Libraries/Types/CodegenTypes';
 import { type HostComponent } from 'react-native';
 
-export interface NativeProps extends $tsflower_subst$RN$ViewProps {
-  mode?: WithDefault<'padding' | 'margin', 'padding'>;
+export type NativeProps = {
+  ...$tsflower_subst$RN$ViewProps,
+  mode?: WithDefault<'padding' | 'margin', 'padding'>,
   edges?: Readonly<{
     top: string,
     right: string,
     bottom: string,
     left: string,
     ...
-  }>;
-}
+  }>,
+  ...
+};
 
 declare var _default: HostComponent<NativeProps>;
 export default _default;
-- 
2.32.0

